File: code/ex404.java
// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public int sumOfLeftLeaves(TreeNode root) {
        if (root == null || (root.left == null && root.right == null))    return 0;
        return pre_order(root, null, 0);
    }

    public int pre_order(TreeNode root, TreeNode lastRoot, int sum) {
        if (root.left != null)
            sum = pre_order(root.left, root, sum);
        
        if (root.right != null)
            sum = pre_order(root.right, root, sum);
        
        if (root.left == null && root.right == null && lastRoot.left == root) {
            sum += root.val;
        }
       
        return sum;
    }
}

File: code/ex404.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int sum = 0;
public:
    int sumOfLeftLeaves(TreeNode* root) {
        if (root == NULL)    return sum;
        dfs(root, 1);
        return sum;
    }
    
private:
    void dfs(TreeNode* root, int dir) {
        if (root->left == NULL && root->right == NULL) {    // Leaves
            if (dir == 0)    sum += root->val;
            return;
        }
        
        if (root->left != NULL)    dfs(root->left, 0);
        if (root->right != NULL)    dfs(root->right, 1);
    }
};